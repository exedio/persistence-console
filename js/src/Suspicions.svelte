<script lang="ts">
  import "@/table-grey.css";
  import { get } from "@/api/api";
  import type { SuspicionsResponse } from "@/api/types";
  import { PromiseTracker } from "@/api/PromiseTracker.svelte";
  import PromiseTrackerReload from "@/api/PromiseTrackerReload.svelte";

  const features = new PromiseTracker(() =>
    get<SuspicionsResponse[]>("suspicions"),
  );
</script>

<table class="grey">
  <caption>
    Suspicions
    <PromiseTrackerReload tracker={features} />
  </caption>
  <thead>
    <tr>
      <th>Feature</th>
      <th>Suspicion</th>
    </tr>
  </thead>
  <tbody>
    {#await features.promise()}
      <tr>
        <td colspan="2" class="empty">fetching data</td>
      </tr>
    {:then features}
      {#each features as feature (feature.type + "." + feature.name)}
        {#each feature.suspicions as suspicion, index}
          <tr>
            {#if index === 0}
              <td rowspan={feature.suspicions.length}
                >{feature.type}.{feature.name}</td
              >
            {/if}
            <td>{suspicion}</td>
          </tr>
        {/each}
      {:else}
        <tr>
          <td colspan="2" class="empty"
            >There are no suspicions for the model.</td
          >
        </tr>
      {/each}
    {:catch error}
      <tr>
        <td colspan="2" class="error">{error.message}</td>
      </tr>
    {/await}
  </tbody>
</table>
