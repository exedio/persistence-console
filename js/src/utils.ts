export function delay<T>(
  call: Promise<T>,
  fail: boolean = false,
  timeout: number = 1000,
): Promise<T> {
  return new Promise((resolve, reject) => {
    setTimeout(() => {
      if (fail) {
        reject(new Error("utils.ts delay failed"));
      } else {
        call.then((r) => resolve(r));
      }
    }, timeout);
  });
}
