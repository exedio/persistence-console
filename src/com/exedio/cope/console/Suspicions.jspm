<%
/*
 * Copyright (C) 2004-2009  exedio GmbH (www.exedio.com)
 *
 * This library is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 2.1 of the License, or (at your option) any later version.
 *
 * This library is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public
 * License along with this library; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
 */

package com.exedio.cope.console;

import com.exedio.cope.Feature;
import java.util.Collection;
import java.util.Iterator;
import java.util.Map;

final class Suspicions_Jspm
{
	@SuppressWarnings("StaticMethodOnlyUsedInOneClass") // ok for jspm
	static void writeBody(
			final Out out,
			final Map<Feature,Collection<String>> features)
	{
		if(features.isEmpty())
		{
			%>
			<table>
				<tr>
					<td class="text">
						There are no suspicions in this model.
					</td>
				</tr>
			</table><%
			return;
		}

		%>
		<table>
			<tr>
				<th>Feature</th>
				<th>Suspicion</th>
			</tr><%
			for(final Map.Entry<Feature,Collection<String>> entry : features.entrySet())
			{
				final Feature feature = entry.getKey();
				final Collection<String> suspicions = entry.getValue();
				final Iterator<String> suspicionsIterator = suspicions.iterator();
				final int suspicionsSize = suspicions.size();
			%>
			<tr>
				<td class="text" rowspan="<%=suspicionsSize%>"><%=feature.getID()%> (<%=feature.getClass()%>)</td>
				<td class="text"><%=suspicionsIterator.next()%></td>
			</tr><%

			while(suspicionsIterator.hasNext())
			{
			%>
			<tr>
				<td class="text"><%=suspicionsIterator.next()%></td>
			</tr><%
			}
			}
			%>
		</table><%
	}

	private Suspicions_Jspm()
	{
		// prevent instantiation
	}
}%>
