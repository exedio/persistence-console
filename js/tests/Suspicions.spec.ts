import { mount } from "svelte";
import { default as Suspicions } from "@/Suspicions.svelte";
import { mockFetch, responseFailure, responseSuccess } from "@t/mockApi";
import { flushPromises, formatHtml } from "@t/utils";
import { expect } from "vitest";
import type { SuspicionsResponse } from "@/api/types";

describe("Suspicions", () => {
  it("should render an empty table", async () => {
    {
      const mock = mockFetch();
      mock.mockResolvedValueOnce(
        responseSuccess([] satisfies SuspicionsResponse[]),
      );
      await mountComponent();
      expect(mock).toHaveBeenCalledExactlyOnceWith("/myApiPath/suspicions");
    }
    expect(await formatHtml(document.body)).toMatchSnapshot();
  });

  it("should render a non-empty table", async () => {
    {
      const mock = mockFetch();
      mock.mockResolvedValueOnce(
        responseSuccess([
          {
            type: "myOnlyType",
            name: "myOnlyName",
            suspicions: ["myOnlySuspicion"],
          },
          {
            type: "myMultiType",
            name: "myMultiName",
            suspicions: ["myFirstSuspicion", "mySecondSuspicion"],
          },
        ] satisfies SuspicionsResponse[]),
      );
      await mountComponent();
      expect(mock).toHaveBeenCalledExactlyOnceWith("/myApiPath/suspicions");
    }
    expect(await formatHtml(document.body)).toMatchSnapshot();

    {
      const mock = mockFetch();
      mock.mockResolvedValueOnce(
        responseSuccess([
          {
            type: "myOnlyTypeReloaded",
            name: "myOnlyNameReloaded",
            suspicions: ["myOnlySuspicionReloaded"],
          },
        ] satisfies SuspicionsResponse[]),
      );
      (document.querySelectorAll(".reload").item(0) as HTMLElement).click();
      await flushPromises();
      expect(mock).toHaveBeenCalledExactlyOnceWith("/myApiPath/suspicions");
    }
    expect(await formatHtml(document.body)).toMatchSnapshot();
  });

  it("should render an error message", async () => {
    {
      const mock = mockFetch();
      mock.mockResolvedValueOnce(responseFailure("myError"));
      await mountComponent();
      expect(mock).toHaveBeenCalledExactlyOnceWith("/myApiPath/suspicions");
    }
    expect(await formatHtml(document.body)).toMatchSnapshot();
  });
});

async function mountComponent() {
  mount(Suspicions, { target: document.body }); // `document` exists because of jsdom
  await flushPromises();
}
