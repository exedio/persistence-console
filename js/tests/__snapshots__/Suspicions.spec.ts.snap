// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Suspicions > should render a non-empty table 1`] = `
"<table class="grey">
  <caption>
    Suspicions
    <button class="reload svelte-dizlwz">ðŸ—˜</button
    ><!---->
  </caption>
  <thead>
    <tr>
      <th>Feature</th>
      <th>Suspicion</th>
    </tr>
  </thead>
  <tbody>
    <!----><!---->
    <tr>
      <td rowspan="1">myOnlyType.myOnlyName</td>
      <!---->
      <td>myOnlySuspicion</td>
    </tr>
    <!---->
    <tr>
      <td rowspan="2">myMultiType.myMultiName</td>
      <!---->
      <td>myFirstSuspicion</td>
    </tr>
    <tr>
      <!---->
      <td>mySecondSuspicion</td>
    </tr>
    <!---->
  </tbody>
</table>
"
`;

exports[`Suspicions > should render a non-empty table 2`] = `
"<table class="grey">
  <caption>
    Suspicions
    <button class="reload svelte-dizlwz">ðŸ—˜</button
    ><!---->
  </caption>
  <thead>
    <tr>
      <th>Feature</th>
      <th>Suspicion</th>
    </tr>
  </thead>
  <tbody>
    <!----><!---->
    <tr>
      <td rowspan="2">myOnlyType.myOnlyName</td>
      <!---->
      <td>myOnlySuspicionChanged</td>
    </tr>
    <tr>
      <!---->
      <td>myOnlySuspicionChangedOther</td>
    </tr>
    <!---->
    <tr>
      <td rowspan="1">myOnlyTypeReloaded.myOnlyNameReloaded</td>
      <!---->
      <td>myOnlySuspicionReloaded</td>
    </tr>
    <!---->
  </tbody>
</table>
"
`;

exports[`Suspicions > should render an empty table 1`] = `
"<table class="grey">
  <caption>
    Suspicions
    <button class="reload svelte-dizlwz">ðŸ—˜</button
    ><!---->
  </caption>
  <thead>
    <tr>
      <th>Feature</th>
      <th>Suspicion</th>
    </tr>
  </thead>
  <tbody>
    <!---->
    <tr>
      <td colspan="2" class="empty">There are no suspicions for the model.</td>
    </tr>
    <!---->
  </tbody>
</table>
"
`;

exports[`Suspicions > should render an error message 1`] = `
"<table class="grey">
  <caption>
    Suspicions
    <button class="reload svelte-dizlwz">ðŸ—˜</button
    ><!---->
  </caption>
  <thead>
    <tr>
      <th>Feature</th>
      <th>Suspicion</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td colspan="2" class="error">suspicions: response code 400 myError</td>
    </tr>
    <!---->
  </tbody>
</table>
"
`;
